src/core/ext/transport/binder/wire_format/wire_writer.cc

```diff
@@ -30,60 +30,126 @@ namespace grpc_binder {
+}
+
+const int64_t WireWriterImpl::kBlockSize = 16 * 1024;
+const int64_t WireWriterImpl::kFlowControlWindowSize = 128 * 1024;
+
 absl::Status WireWriterImpl::RpcCall(const Transaction& tx) {
   // TODO(mingcl): check tx_code <= last call id
   grpc_core::MutexLock lock(&mu_);
   GPR_ASSERT(tx.GetTxCode() >= kFirstCallId);
-  RETURN_IF_ERROR(binder_->PrepareTransaction());
-  WritableParcel* parcel = binder_->GetWritableParcel();
-  {
-    //  fill parcel
+  int& seq = seq_num_[tx.GetTxCode()];
+  if ((tx.GetFlags() & kFlagMessageData) == 0 ||
+      tx.GetMessageData().size() <= kBlockSize) {
+    // Fast path: send data in one transaction.
+    RETURN_IF_ERROR(binder_->PrepareTransaction());
+    WritableParcel* parcel = binder_->GetWritableParcel();
     RETURN_IF_ERROR(parcel->WriteInt32(tx.GetFlags()));
-    RETURN_IF_ERROR(parcel->WriteInt32(tx.GetSeqNum()));
+    RETURN_IF_ERROR(parcel->WriteInt32(seq++));
     if (tx.GetFlags() & kFlagPrefix) {
-      // prefix set
-      if (tx.IsClient()) {
-        // Only client sends method ref.
-        RETURN_IF_ERROR(parcel->WriteString(tx.GetMethodRef()));
-      }
-      RETURN_IF_ERROR(parcel->WriteInt32(tx.GetPrefixMetadata().size()));
-      for (const auto& md : tx.GetPrefixMetadata()) {
-        RETURN_IF_ERROR(parcel->WriteByteArrayWithLength(md.first));
-        RETURN_IF_ERROR(parcel->WriteByteArrayWithLength(md.second));
-      }
+      RETURN_IF_ERROR(WriteInitialMetadata(tx, parcel));
     }
     if (tx.GetFlags() & kFlagMessageData) {
       RETURN_IF_ERROR(parcel->WriteByteArrayWithLength(tx.GetMessageData()));
     }
     if (tx.GetFlags() & kFlagSuffix) {
-      if (tx.IsServer()) {
-        if (tx.GetFlags() & kFlagStatusDescription) {
-          RETURN_IF_ERROR(parcel->WriteString(tx.GetStatusDesc()));
-        }
-        RETURN_IF_ERROR(parcel->WriteInt32(tx.GetSuffixMetadata().size()));
-        for (const auto& md : tx.GetSuffixMetadata()) {
-          RETURN_IF_ERROR(parcel->WriteByteArrayWithLength(md.first));
-          RETURN_IF_ERROR(parcel->WriteByteArrayWithLength(md.second));
-        }
-      } else {
-        // client suffix currently is always empty according to the wireformat
-        if (!tx.GetSuffixMetadata().empty()) {
-          gpr_log(GPR_ERROR, "Got non-empty suffix metadata from client.");
-        }
+      RETURN_IF_ERROR(WriteTrailingMetadata(tx, parcel));
+    }
+    // FIXME(waynetu): Construct BinderTransportTxCode from an arbitrary integer
+    // is an undefined behavior.
+    return binder_->Transact(BinderTransportTxCode(tx.GetTxCode()));
+  }
+  // Slow path: the message data is too large to fit in one transaction.
+  int original_flags = tx.GetFlags();
+  GPR_ASSERT(original_flags & kFlagMessageData);
+  absl::string_view data = tx.GetMessageData();
+  size_t ptr = 0;
+  while (ptr < data.size()) {
+    while (num_outgoing_bytes_ >=
+           num_acknowledged_bytes_ + kFlowControlWindowSize) {
+      cv_.Wait(&mu_);
```

Comment:
Probably shouldn't block here, it would be pretty hard to debug if we don't receive ACK from the other end for some reason (since the program will just hang).

We probably want a helper class or something to handle chunking so that flow control logic (which can be orthogonal to other logic) don't make `WireWriterImpl` become too complex

If that's too much to implement in the same PR, probably add a 10ms deadline to `cv_.Wait()` (and log ERROR if deadline exceeded) and add a TODO

Reply:
How about having a new class like `FlowControl` to take care of the flow control mechanism. The transport now holds a `FlowControl` (which internally holds a `WieWtier` to deal with the wire format) and the transport sends RPC call through `FlowControl` instead of `WireWriter`.

What do you think?

Reply:
I think that's probably too much abstraction? It should be possible to implement the flow control logic within WireWriterImpl and make the code readable? Probably introduce some kind of queue as member variable and some member functions

Reply:
Done.